tests/unit/s2n_self_talk_shutdown_test.c

```diff
@@ -0,0 +1,259 @@
+        EXPECT_SUCCESS(s2n_io_pair_init_non_blocking(&io_pair));
+        EXPECT_SUCCESS(s2n_connection_set_io_pair(client_conn, &io_pair));
+        EXPECT_SUCCESS(s2n_connection_set_io_pair(server_conn, &io_pair));
+
+        /* Verify state prior to alert */
+        EXPECT_FALSE(server_conn->close_notify_received);
+        EXPECT_FALSE(client_conn->close_notify_received);
+
+        s2n_blocked_status server_blocked;
+        s2n_blocked_status client_blocked;
+
+        /* Verify successful shutdown. Server initiated. */
+        EXPECT_FAILURE_WITH_ERRNO(s2n_shutdown(client_conn, &client_blocked), S2N_ERR_IO_BLOCKED);
+        EXPECT_SUCCESS(s2n_shutdown(server_conn, &server_blocked));
+        EXPECT_SUCCESS(s2n_shutdown(client_conn, &client_blocked));
+
+        /* Verify state after alert */
+        EXPECT_SUCCESS(server_conn->close_notify_received);
+        EXPECT_SUCCESS(client_conn->close_notify_received);
+
+        /* Cleanup */
+        EXPECT_SUCCESS(s2n_connection_free(server_conn));
+        EXPECT_SUCCESS(s2n_connection_free(client_conn));
+        EXPECT_SUCCESS(s2n_io_pair_close(&io_pair));
+        EXPECT_SUCCESS(s2n_cert_chain_and_key_free(chain_and_key));
+        EXPECT_SUCCESS(s2n_config_free(config));
+    }
+
+    /* Verify successful shutdown. Client and Server initiated. */
+    {
+        /* Setup connections */
+        struct s2n_connection *client_conn, *server_conn;
+        EXPECT_NOT_NULL(client_conn = s2n_connection_new(S2N_CLIENT));
+        EXPECT_NOT_NULL(server_conn = s2n_connection_new(S2N_SERVER));
+
+        /* Setup config */
+        struct s2n_cert_chain_and_key *chain_and_key;
+        EXPECT_SUCCESS(s2n_test_cert_chain_and_key_new(&chain_and_key,
+                S2N_DEFAULT_ECDSA_TEST_CERT_CHAIN, S2N_DEFAULT_ECDSA_TEST_PRIVATE_KEY));
+        struct s2n_config *config;
+        EXPECT_NOT_NULL(config = s2n_config_new());
+        EXPECT_SUCCESS(s2n_config_set_cipher_preferences(config, "default_tls13"));
+        EXPECT_SUCCESS(s2n_config_set_unsafe_for_testing(config));
+        EXPECT_SUCCESS(s2n_config_add_cert_chain_and_key_to_store(config, chain_and_key));
+        EXPECT_SUCCESS(s2n_connection_set_config(client_conn, config));
+        EXPECT_SUCCESS(s2n_connection_set_config(server_conn, config));
+
+        /* Create nonblocking pipes */
+        struct s2n_test_io_pair io_pair;
+        EXPECT_SUCCESS(s2n_io_pair_init_non_blocking(&io_pair));
+        EXPECT_SUCCESS(s2n_connection_set_io_pair(client_conn, &io_pair));
+        EXPECT_SUCCESS(s2n_connection_set_io_pair(server_conn, &io_pair));
+
+        /* Verify state prior to alert */
+        EXPECT_FALSE(server_conn->close_notify_received);
+        EXPECT_FALSE(client_conn->close_notify_received);
+
+        /* Verify state when we dont call shutdown. Not a very interesting test case. */
+        EXPECT_FALSE(server_conn->close_notify_received);
+        EXPECT_FALSE(client_conn->close_notify_received);
```

Comment:
I don't see a test here?

For the stated use case, wouldn't you need both the client and server to write the close notify before either reads it? I would suggest checking out [stuffer io](https://github.com/aws/s2n-tls/blob/main/tests/testlib/s2n_testlib.h#L39), bc you could potentially copy into the inputs after the connections write to the outputs. That may be straying out of self-talk territory and back into unit test territory, I I don't think S2N is capable of reliably setting up the "both close" situation otherwise.

Reply:
yep this test doesnt make sense and ill be removing it.